packages/cli/src/handlers/dbt/compile.ts

```diff
@@ -69,6 +72,103 @@ export const dbtCompile = async (options: DbtCompileOptions) => {
     }
 };
 
+// TODO: Move this to somewhere appropriate
+type UnrenderedConfig =
+    | {
+          meta?: {
+              joins?: Array<{ join: string }>;
+          };
+      }
+    | undefined;
+
+const recursivelyGetJoinedModelNames = (
+    modelNode: DbtModelNode,
+    allModelNodes: DbtModelNode[],
+): string[] => {
+    const joinedModelNames = (
+        modelNode.unrendered_config as UnrenderedConfig
+    )?.meta?.joins?.map((j) => j.join);
+
+    if (!joinedModelNames) {
+        return [];
+    }
+
+    const joinedModelNodes = allModelNodes.filter((model) =>
+        joinedModelNames.includes(model.name),
+    );
+
+    return joinedModelNodes.reduce<string[]>(
+        (acc, model) => [
+            ...acc,
+            model.name,
+            ...recursivelyGetJoinedModelNames(model, allModelNodes),
+        ],
+        [],
+    );
+};
+
+export const dbtCompileNecessaryModels = async (
+    loadManifestOpts: LoadManifestArgs,
+    options: DbtCompileOptions,
+) => {
+    // If no models are explicitly selected or excluded, compile all models
+    if (!options.select && !options.exclude) {
+        await dbtCompile(options);
+        return;
+    }
+
+    const manifest = await loadManifest(loadManifestOpts);
+    const manifestModels = getModelsFromManifest(manifest);
+
+    const selectedModelNames = (
+        options.select ?? manifestModels.map((model) => model.name)
+    ).filter((modelName) => !options.exclude?.includes(modelName));
+
+    const selectedModelNodes = manifestModels.filter((model) =>
+        selectedModelNames.includes(model.name),
+    );
+
+    // Selected models and their joined models
+    const modelNamesToCompile = new Set(
+        selectedModelNodes.reduce<string[]>((acc, model) => {
+            const joinedModelNames = recursivelyGetJoinedModelNames(
+                model,
+                manifestModels,
+            );
+            return [...acc, model.name, ...(joinedModelNames ?? [])];
+        }, []),
+    );
+
+    await dbtCompile({
+        ...options,
+        select: Array.from(modelNamesToCompile),
+    });
+};
+
+export const dbtParse = async (
+    dbtVersion: string,
+    opts: Pick<
+        DbtCompileOptions,
+        'profilesDir' | 'projectDir' | 'profile' | 'target'
+    >,
+) => {
+    try {
+        const args = optionsToArgs(opts);
+
+        if (dbtVersion.startsWith('1.3.') || dbtVersion.startsWith('1.4.')) {
+            args.push('--write-manifest');
+        }
```

Comment:
https://docs.getdbt.com/reference/commands/parse

From DBT Docs: "Starting in v1.5, dbt parse will write or return a [manifest](https://docs.getdbt.com/reference/artifacts/manifest-json), enabling you to introspect dbt's understanding of all the resources in your project."